src/cargo/ops/cargo_add/mod.rs

```diff
@@ -364,7 +364,24 @@ fn resolve_dependency(
     };
     selected_dep = populate_dependency(selected_dep, arg);
 
-    let old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    let mut old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    if old_dep.is_none() && selected_dep.source().is_none() && selected_dep.rename().is_none() {
```

Comment:
Why the `selected_dep` checks?

Reply:
`selected_dep.source().is_none()` to try the mutations only on when looking for registry dependencies and `selected_dep.rename().is_none()` to not try permutations when adding a renamed dependency.

Reply:
Could you expand on the rename case?

Reply:
I changed the implementation - I believe this discussion is not relevant any more.

Reply:
Please squash when you completely change directions in a PR.  It can be confusing to track what changed or didn't change between commits.

Reply:
So while the code is changed, the concepts from this discussion are still there
```rust
    if let Some(rename) = dependency.rename() {
        return lookup(rename);
    }
```
```rust
            if !matches!(dep.source, Some(Source::Registry(_))) {
                continue;
            }
```

The motivation for these checks is not obvious.  Could you explain these?

Also, if there aren't tests that would break with their removal, could you add them?

Reply:
I added comments that explain the conditions.

When I remove the first condition these tests fail:

cargo_add::change_rename_target::case
cargo_add::invalid_key_rename_inherit_dependency::case
cargo_add::overwrite_rename_with_rename_noop::case

so that is covered.

That is not the case for the second condition, so I added a case to the test suite.